Email Editor - Fix Email Editor Permissions for Non-Admin Users

triple0t opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->

Closes #55909 

This PR addresses permission-related issues in the Email Editor that were affecting non-admin WordPress users, particularly those with the Editor role. The changes ensure proper permission checks are in place and that the editor functions correctly for users with limited permissions.

#### Key Changes
1. **Permission-Based Template Editing**
   - Added role-based permission checks for template editing
   - Modified template retrieval to respect user capabilities
   - Implemented proper context ('view' vs 'edit') based on user permissions

2. **Global Styles Access Control**
   - Added user permission checks for the Global Styles sidebar
   - Modified Global Styles access to respect user capabilities
   - Updated API endpoints to use appropriate context based on user roles

3. **UI/UX Improvements**
   - Added informative messages for users without template editing permissions
   - Conditionally render template editing options based on user capabilities
   - Improved error handling for permission-restricted actions

4. **Code Refactoring**
   - Refactored template editing permissions and template retrieval logic
   - Streamlined permission checks across components
   - Enhanced code organization and maintainability

### Screenshots or screen recordings:

<!-- If this PR includes UI changes, please provide screenshots or a screen recording for clarity. -->
<!-- This section can be removed if not applicable. -->

| Before | After |
| ------ | ----- |
|  ![Screenshot 2025-03-10 at 2 42 56 PM](https://github.com/user-attachments/assets/8b51adc0-4278-49c2-a1a4-31c216cdd2c6) | ![Screenshot 2025-03-10 at 2 44 19 PM](https://github.com/user-attachments/assets/7f65e952-d48e-4cdb-8989-bbc0b43f2a53)  |


<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Log in as a user with Editor role
2. Navigate to the Email Editor
3. Verify that appropriate UI elements are shown/hidden based on permissions i.e. Global styles icon. edit template
4. Confirm that template viewing works correctly
5. Ensure Global Styles sidebar respects user permissions i.e The Global styles should not show for non-admin users
6. Verify that appropriate error messages are shown for restricted actions

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [x] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->
Fixes the issue where the Email Editor wasn't working properly for users with the Editor role.
</details>

triple0t committed:
Email Editor: Modify template retrieval to return template titles

This fixes the error `PHP Warning:  array_flip(): Can only flip string and integer values, entry skipped in /var/www/html/wp-admin/includes/theme.php`

triple0t committed:
Email Editor: Add role-based template editing permissions

- Add permission check for template editing using `canUser` selector
- Implement different modals for admin and non-admin users when attempting to edit template blocks
- Conditionally render edit template menu item based on user permissions

triple0t committed:
Email Editor: Add user permission check for global styles sidebar

- Add `userCanEditGlobalStyles` check using `useSelect` hook
- Conditionally render styles sidebar based on user permissions
- Update selectors to support retrieving global styles with proper context
- Modify REST API routes to use 'view' context for global styles

triple0t committed:
Email Editor: Refactor template editing permissions and template retrieval

- Replace `canCreateTemplates` with `canUpdateTemplates` using a new store selector
- Remove direct dependency on `@wordpress/core-data` in components
- Add `regularizedGetEntityRecord` helper to normalize template record retrieval because of the difference between getEditedEntityRecord and getEntityRecord
- Implement `getTemplate` function to handle template fetching with proper permissions
- Update selectors to use 'view' context for template and global styles records

triple0t committed:
Email Editor: Add change log file for fixing permissions for editor role

Resolve an issue preventing the Email Editor from loading Global Styles for non-admin WordPress users by ensuring correct permission checks are implemented.

lysyjan reviewed (changes requested):
@triple0t When I was testing this PR, I had a problem with editing the template as the admin user.

![Screenshot on 2025-03-10 at 18-21-50](https://github.com/user-attachments/assets/2e064223-7c88-4d99-b924-b0c9ff79a4c4)
![Screenshot on 2025-03-10 at 18-23-58](https://github.com/user-attachments/assets/a382210e-34ac-42c3-a156-be480f785e26)

Please can you check it?

triple0t committed:
Email Editor: Fix template editing permission check

Change the permission check from 'update' to 'create' for template editing to ensure correct user access validation

lysyjan reviewed (approved):
I have tested it and it works as I expected. 
I tried:
- creating email as the editor and the admin
- editing emails as the editor and the admin
- verified that the editor cannot modify styles and templates
- verified that the admin can modify styles and templates
- both roles can see expected dialog

The code also looks good.

<img width="307" alt="Screenshot 2025-03-12 at 11 28 11" src="https://github.com/user-attachments/assets/828a0fb5-b5f4-49ab-a5b9-ac941a18450b" />
<img width="1127" alt="Screenshot 2025-03-12 at 11 27 56" src="https://github.com/user-attachments/assets/80327355-84f8-4779-b2fe-b608f4d96b7c" />
<img width="1118" alt="Screenshot 2025-03-12 at 11 25 26" src="https://github.com/user-attachments/assets/ef684418-d886-4a3f-861a-039a20ac93d2" />
